pkg/agent/controller/supportbundlecollection/support_bundle_controller.go

```diff
@@ -0,0 +1,417 @@
+// Copyright 2022 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package supportbundlecollection
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"path"
+	"reflect"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/pkg/sftp"
+	"github.com/spf13/afero"
+	"golang.org/x/crypto/ssh"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/exec"
+
+	"antrea.io/antrea/pkg/agent"
+	"antrea.io/antrea/pkg/agent/config"
+	agentquerier "antrea.io/antrea/pkg/agent/querier"
+	"antrea.io/antrea/pkg/apis/controlplane"
+	cpv1b2 "antrea.io/antrea/pkg/apis/controlplane/v1beta2"
+	"antrea.io/antrea/pkg/apis/crd/v1alpha1"
+	"antrea.io/antrea/pkg/ovs/ovsctl"
+	"antrea.io/antrea/pkg/querier"
+	"antrea.io/antrea/pkg/support"
+	"antrea.io/antrea/pkg/util/compress"
+)
+
+type ProtocolType string
+
+const (
+	SFTP ProtocolType = "SFTP"
+
+	controllerName = "SupportBundleCollectionController"
+	defaultWorkers = 1
+
+	UPLOAD_TO_FILESERVER_TRIES       = 5
+	UPLOAD_TO_FILESERVER_RETRY_DELAY = 5 * time.Second
+)
+
+var (
+	emptyWatch      = watch.NewEmptyWatch()
+	defaultFS       = afero.NewOsFs()
+	defaultExecutor = exec.New()
+	// Declared as variable for testing.
+	newAgentDumper = support.NewAgentDumper
+
+	getUploader          = getUploaderByProtocol
+	sftpUploaderInstance *sftpUploader
```

Comment:
could this be stored as a member variable of SupportBundleController? If yes, I would suggest to do it as global variable is not unit test friendly and not good for module encapsulation.

Reply:
Actually, I use it as a global variable in order to let it be used by getUploaderByProtocol, so that I can mock getUploader in the unit test. If it is a member variable, I think it will be hard for us to mock this one in unit test.

Reply:
Seems we are thinking the same purpose but opposite approach. I think member variable is more friendly to unit test, example as below:
```
type SupportBundleController struct {
	...

	sftpUploader uploader
}

func NewSupportBundleController(nodeName string,
	nodeType config.NodeType,
	namespace string,
	antreaClientGetter agent.AntreaClientProvider,
	ovsCtlClient ovsctl.OVSCtlClient,
	aq agentquerier.AgentQuerier,
	npq querier.AgentNetworkPolicyInfoQuerier,
	v4Enabled,
	v6Enabled bool) *SupportBundleController {
	c := &SupportBundleController{
		...
		sftpUploader: &sftpUploader{},
	}
	return c
}

func (c *SupportBundleController) getUploaderByProtocol(protocol ProtocolType) (uploader, error) {
	if protocol == sftpProtocol {
		return c.sftpUploader, nil
	}
	return nil, fmt.Errorf("unsupported protocol %s", protocol)
}

func newFakeController(t *testing.T, nodeType config.NodeType, fakeSFTPUploader uploader) (*fakeController, *fakeversioned.Clientset) {
	controller := gomock.NewController(t)
	clientset := &fakeversioned.Clientset{}
	supportBundleController := NewSupportBundleController("vm1", nodeType, "vm-ns", &antreaClientGetter{clientset}, nil,
		nil, nil, true, true)
        supportBundleController.sftpUploader = fakeSFTPUploader
	return &fakeController{
		SupportBundleController: supportBundleController,
		mockController:          controller,
	}, clientset
}
```
The two global variables are not needed and unit test just needs to inject a fake uploader to controller and no need to worry about cleanup.